### DISCUSSION TITLE
Fix latency with PulseAudio and simultaneous output

### DISCUSSION THREAD

**Discussion by HorstBaerbel** [snapcast/snapcast#868] [Q&A] [answered]
I have a setup here with a Ryzen 2400G PC with Ubuntu Desktop 20.04.2 and Kodi 19 installed. I installed the snapcast package and configured it:
```
sampleformat = 44100:16:2
buffer = 200
```
PulseAudio is configured to have a combined sink, simultaneously playing back on the hardware speaker and snapcast server using:
```
### Set up snapcast server sink
load-module module-pipe-sink file=/tmp/snapfifo sink_name=Snapcast format=s16le rate=44100
update-sink-proplist Snapcast device.description=Snapcast

### Set up combined sink
load-module module-combine-sink sink_name=combined slaves=<HW_CARD>.<HW_SINK>,Snapcast
update-sink-proplist combined device.description="All devices simultaneously"
set-default-sink combined
```
Note that I didn't need to set `PULSE_LATENCY_MSEC=60` in /etc/environment.

This works okish. The sound from the system and Kodi is played back on the speaker and via snapcast / the snapdroid Android app on a mobile phone.
BUT the latency is pretty audible and fluctuating. I tried setting the latency in the Android app to 0ms, 50ms or 100ms, I also tried setting the PulseAudio device latency for the output device to 0ms, 100ms, 150ms using:
`pacmd set-port-latency-offset <HW_CARD> <HW_PROFILE> 150000`
but I'm not getting synchronous audio playback. The weird thing is that when I set the PulseAudio output latency to 100ms, it gets audibly better, but once I set higher values (or wait a bit, I'm not sure), it becomes worse again and no matter what I set after that changes nothing... Strange.

What can I do to fix the latency? Would the Kodi Addon help me in any way? Might Ubuntu 21.04 improve this in any way?

**Comment by vjdw** [comment, marked as answer]
Output from snapclient (like on your Android phone) is always going to have a variable delay because that's how snapcast works. The audio output on each snapcast client is matched to whichever client has the highest latency, and there will always be some latency because of buffering.

The audio you're sending straight to your local sink will always get played with relatively low latency.

The best fix might be to run snapclient on the server itself instead of sending anything directly to the hardware sink directly (i.e. don't use combined sink, but listen in to the snapcast sink using snapclient).

**Reply by HorstBaerbel** [reply]
Is there an option in snapcast server to do that directly, e.g. use snapserver as a pulseaudio "filter / effect" device?

**Reply by HorstBaerbel** [reply]
And if I have that working should the playback then be (more or less) simultaneous automatically?

**Reply by vjdw** [reply]
Not sure what you mean about the "filter / effect" thing but it doesn't sound like anything I've seen in snapserver.

You'd have to install snapclient, just like you've already installed snapserver. When you've done that it should start playing in sync with the Android device.

**Reply by HorstBaerbel** [reply]
I removed the combined sink, installed snapclient and edited /etc/default/snapclient to:
`SNAPCLIENT_OPTS="--player pulse --host 127.0.0.1"`

This works, but at bootup I get the some service error messages:
Exception in Controller::worker(): Can't open PCM device: Device or resource busy
and a ubuntu error popup. Is there a way to get rid of that?

**Reply by idpnd** [reply]
Since the delay is variable, this could be approximated using a manual audio/video offset. If a value for the offset could be extracted from snapserver, this could be used to script an adaptive a/v offset..
